lib/ruby-debug-ide/xml_printer.rb

```diff
@@ -385,28 +385,33 @@ def max_compact_name_size
     end
 
     def inspect_with_allocation_control(slice, memory_limit)
-      curr_thread = Thread.current
-      
-      start_alloc_size = ObjectSpace.memsize_of_all
-      
-      trace = TracePoint.new(:c_call, :call) do |tp|
-        curr_alloc_size = ObjectSpace.memsize_of_all
-        
-        if(curr_alloc_size - start_alloc_size > 1e6 * memory_limit)
-          
-          trace.disable
-          curr_thread.raise MemoryLimitError, "Out of memory: evaluation of inspect took more than #{memory_limit}mb." if curr_thread.alive?
-        end
+      if Thread.respond_to?(:critical) and Thread.critical
+        raise ThreadError, "memory_limit within critical session"      
       end
+      
+      curr_thread = Thread.current
 
-      trace.enable
-      result = slice.inspect
-      trace.disable
-      result 
+      result = nil
+      inspect_thread = DebugThread.start {
+        start_alloc_size = ObjectSpace.memsize_of_all
+        trace = TracePoint.new(:c_call, :call) do |tp|
+          curr_alloc_size = ObjectSpace.memsize_of_all
+          start_alloc_size = curr_alloc_size if (curr_alloc_size < start_alloc_size)
+
+          if(curr_alloc_size - start_alloc_size > 1e6 * memory_limit)
+            curr_thread.raise MemoryLimitError, "Out of memory: evaluation of inspect took more than #{memory_limit}mb. \n#{caller.map{|l| "\t#{l}"}.join("\n")}"
+            inspect_thread.kill
```

Comment:
I'd rather move this kill to line 410, after join. It seems strange to me to kill the thread from this thread itself.